### PR TITLE
Shay/add pedestrians

### PR DESCRIPTION
**Pull request by JimmyLeeJonesOG**
Adding pedestrians at road cells with customizable probability for them to be generated. The pedestrians are a copy of an original prefabed pedestrian. It's not perfect, especially around the edges, but it's something

### CODE REVIEW

File: Assets/Code/Behaviours/GenerateCity.cs [hunk 1]
```diff
@@ -1,23 +1,40 @@
...
 using System.Collections.Generic;
 using UnityEngine;
 using static Map;
+using Random = UnityEngine.Random;
 
 public class GenerateCity : MonoBehaviour
 {
-    const float cellSize = 10f;
+    const float cellSize = 10f;
+    private const int Rows = 20;
```
Comment (yossistarz) [inline Assets/Code/Behaviours/GenerateCity.cs:10]: Thanks 👍

File: Assets/Code/Behaviours/GenerateCity.cs [hunk 2]
```diff
@@ -94,6 +117,40 @@ void Start()
...
+    public MapCell GetCellFromPosition(Vector3 position)
+    {
+        var relativePosition = position - bottomLeftCornerPosition;
+        var row = Mathf.FloorToInt(relativePosition.x) / Mathf.RoundToInt(cellSize);
+        var col = Mathf.FloorToInt(relativePosition.z) / Mathf.RoundToInt(cellSize);
+        try
+        {
+            return _map.GetMapCell(row, col);
+        }
+        catch (Exception)
```
Comment (yossistarz) [inline Assets/Code/Behaviours/GenerateCity.cs]: Maybe we should add a check in the GetMapCell for the row and the col.
The exceptions are very slow mechanism, this should slow down the game quite a bit.

File: Assets/Code/Behaviours/GenerateCity.cs [hunk 3]
```diff
@@ -94,6 +117,40 @@ void Start()
...
+    {
+        return (position.x < bottomLeftCornerPosition.x || position.x > bottomLeftCornerPosition.x + Rows * cellSize ||
+                position.z < bottomLeftCornerPosition.z || position.z > bottomLeftCornerPosition.x + Cols * cellSize);
+    }
+
+    void Update()
+    {
+        foreach (var pedestrian in _pedestrians)
+        {
+            var positionOfLookingAhead = pedestrian.transform.position + pedestrian.Direction * _lookAheadDistanceMagnitude;
```
Comment (yossistarz) [inline Assets/Code/Behaviours/GenerateCity.cs:139]: This is cool.
I do have an idea for next version.
We could use the Cell's ConnectionPoints (bad name, i know 😅) which represents the relative path (from 0,0 to 1,1) to the center of the cell that this cell containes.

So for example, on a cell that has only 1 direction, you will have two connections points.
one from the center to the first edge, and the other from the center to the other edge.
This helps you know how to move in this cell.
So you could know the direction you need to walk if the cell is a line or it is a left/right turn, or even if it is an intersection which means you will have 3 points or 4 points (depends on how many roads are connected together.
So you will know when and how to turn.

(Not needed now :) )

File: Assets/Code/Behaviours/GenerateCity.cs [hunk 4]
```diff
@@ -94,6 +117,40 @@ void Start()
...
+            var cellOfLookingForward = IsPositionOutOfBounds(positionOfLookingAhead) ? null : GetCellFromPosition(positionOfLookingAhead);
```
Comment (yossistarz) [inline Assets/Code/Behaviours/GenerateCity.cs:146]: Also i have two suggestions about the update function.
- You can either have 1 pedestrian that you can create like a template (the same way i did with the roads), and than have a separate behavior which is bound to the GenerateCity behavior to get the map (the same why i did with the behavior that puts the car randomly on the map). Than this behavior moves each pedestrian on it's own -> you can create a more richer environment with simpler code that way.
- You can extract this entire code into a different single behavior to handle all the pedestrians, but only them.

I think that we have performance considerations that we will need at address if we take one or the other, but this is really way ahead of time to talk about that 😅.

Comment (yossistarz) [inline Assets/Code/Behaviours/GenerateCity.cs:140]: I have a suggestion about the update function for next iteration maybe.

behavior which is bound to the GenerateCity behavior to get the map (the same why i did with the behavior that puts the car randomly on the map). Than this behavior moves each pedestrian (As you already did) and can check the info on the map, the same way you did here :).

File: Assets/Code/Behaviours/GenerateCity.cs [hunk 5]
```diff
@@ -27,10 +44,16 @@ void Start()
...
-                {
-                    if (cell.MapCellType == Map.MapCellType.Road)
-                    {
-                       
+                {
+                    if (cell.MapCellType == MapCellType.Road)
+                    {
+                        if (Random.Range(0f, 1f) < CreatePedestrainProbability)
+                        {
+                            var newPedestrian = Instantiate(pedestrian, GetNewPedestrianPosition(cell), Quaternion.identity);
```
Comment (yossistarz) [inline Assets/Code/Behaviours/GenerateCity.cs:52]: Also can be done on some other iteration.

But i need to clean up the code in this class (it was created with a lot of things, just to make it work quickly and dirty).
So, i think we should also extract the pedestrian creation into another behavior that has binding to this behavior (same way i did with the car positioning).

File: Assets/Pedestrian.cs [hunk 6]
```diff
@@ -0,0 +1,56 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+using Random = UnityEngine.Random;
+
+public class Pedestrian : MonoBehaviour
+{
+    public float SpeedMagnitude = 1f;
+    public Vector3 Direction;
```
Comment (yossistarz) [inline Assets/Pedestrian.cs:10]: BTW, about the Direction, you can use also the pedestrian.transform.forward to get the direction it is heading 😃.

**Review by yossistarz — APPROVED** [review]
Cool update :).

You are welcome to push the update, but please don't skip the comments 😊.

File: Assets/Code/Behaviours/GenerateCity.cs (hunk 2 above)
Comment (JimmyLeeJonesOG) [reply Assets/Code/Behaviours/GenerateCity.cs]: you're right, moved to GetMapCell

**Pull request merged**